django/forms/widgets.py

```diff
@@ -823,7 +827,10 @@ def get_context(self, name, value, attrs):
         for i, widget in enumerate(self.widgets):
             if input_type is not None:
                 widget.input_type = input_type
-            widget_name = '%s_%s' % (name, i)
+            if self.names:
+                widget_name = self.names and self.names[i] or i
+            else:
+                widget_name = '%s_%s' % (name, i)
```

Comment:
This needs to incorporate `name` somehow else it will break when form prefixes are involved or on top level field names collisions.

Reply:
Hi @charettes I've added the field name and also allowed the underscore to be removed if an empty string is passed in as the widget name. I've also removed the index as names are now keys from a dictionary. 

I've also added the blank widget name option to the tests. 

Remaining question on this part is should we now be using f strings?